src/mlir/passes.md

```diff
@@ -0,0 +1,207 @@
+   (TBAA).
+
+When the result of the analysis is that it *cannot* alias, we're free to do any
+optimisation straight away. For those that *can*, we need further analysis to
+separate them between *must*, *may* and *must-not*.
+
+This is the analysis that tracks provenance (the origin of the references) to
+make sure they come from disjoint memory. These can come from:
+ * local variables, for example via assignments (`var`, `let`), in which case a
+   local analysis is enough.
+ * function arguments, in which case we need to find all callers of the function
+   and separate between aliasing calls and non-aliasing calls. If we only have
+   non-aliasing calls, we can optimise as before. If not, we may choose to create
+   multiple versions of the function and change the original calls to match.
+ * class fields, which could have been modified by a previous step, depending on
+   runtime behaviour, and are therefore very hard to prove either way.
+
+### IR state afterwards
+
+After the alias analysis step, the IR shouldn't change much in structure.
+
+If we found opportunities for optimisation at the MLIR level, then some code may
+have simplified, but not changed in dialect support. For the rest of the alias
+information, we should annotate the operations to carry on TBAA info into the
+LLVM IR and allow LLVM to optimise on its own.
+
+## Dynamic vs static calls
+
+Static calls are cheaper than dynamic calls. The former is represented in MLIR
+and LLVM IR natively and translate to direct jumps into code memory. The latter
+has to find the object's vtable address, the method's offset and do an indirect
+call on that address.
+
+We will generate a function for every method, taking the receiver as the
+first argument. Calls to concrete types are static call on the type's
+function directly and therefore need no additional treatment.
+
+For all the call sites where the target is an interface type, we should insert
+indirect (vtable) calls for every class that is ever cast to an interface type,
+containing only the methods that might be called via an interface.
+
+This minimises the size and complexity of the vtables and improve runtime
+performance avoiding unnecessary indirect calls.
+
+### IR state afterwards
+
+At this stage, all static calls should have moved to a direct `std.call` to
+named (and possibly mangled) names that exist as functions in the IR, with the
+appropriate lexical context (module name, etc).
+
+## Virtual table colouring
+
+Concrete classes implementing the same interface will have similar (if not
+identical) method layout. Concrete implementation of generic functions can
+proliferate concrete types, increasing the number of similar/identical type
+representations.
+
+If each type has its own vtable, we're spending a large amount of read-only
+memory to store them all. This can also hurt caching, if objects are used in hot
+code but are too far apart from each other in the read-only section.
```

Comment:
I have no data to back this up, but I seriously think this is negligeable and is certainly a problem that doesn't need solving now.

Reply:
No one is saying we'll solve it now.